src/codegen/compile.ml

```diff
@@ -1602,12 +1608,29 @@ module Tagged = struct
     assert (!Flags.gc_strategy = Flags.Incremental);
     1l
 
-  (* Note: Post allocation barrier must be applied after initialization *)
+  (* Note: post-allocation barrier must be applied after initialization *)
   let alloc env size tag =
     let name = Printf.sprintf "alloc_size<%d>_tag<%d>" (Int32.to_int size) (Int32.to_int (int_of_tag tag)) in
+    let overflow_mask n =
+      let n = Int32.to_int n in
+      Int32.(logand 0xFFFFl (shift_left minus_one (16 - Numerics.Nat16.(to_int (clz (of_int n)))))) in
+
     Func.share_code0 env name [I32Type] (fun env ->
-      let (set_object, get_object) = new_local env "new_object" in
-      Heap.alloc env size ^^
+      let set_object, get_object = new_local env "new_object" in
+      (if !Flags.gc_strategy <> Flags.Incremental && size < 0x2000l then
+         GC.get_heap_pointer env ^^
+         compile_sub_const 1l ^^ (* skew *)
+         GC.get_heap_pointer env ^^
+         let size_in_bytes = Int32.(mul size Heap.word_size) in
+         compile_add_const size_in_bytes ^^
+         GC.set_heap_pointer env ^^
+         GC.get_heap_pointer env ^^
+         compile_bitand_const (overflow_mask size_in_bytes) ^^
+         G.if0
+           G.nop (* no page crossing *)
+           (Heap.alloc env 0l ^^ G.i Drop) (* ensure that the page is allocated *)
```

Comment:
What is this crosses multiple pages, will the intermediate pages also be allocated? If size > 65535 (in bytes) I think you could cross multiple pages

Reply:
It would commit many pages, but it can't due to the size check above (`< 0x2000` words)

Reply:
Right you are. I was reading my hex wrong. That's 32768 bytes.

Reply:
3f301d74a performs the comparison in bytes, this will work for 8-byte words too.

Reply:
See also https://github.com/dfinity/motoko/pull/4064#issuecomment-1607963648